pypit/core/artracewave.py

```diff
@@ -0,0 +1,1235 @@
+    yspl = np.append(0.0, np.append(ycen, 1.0))
+    zspl = np.zeros((msarc.shape[1], np.sum(aduse) + 2))
+    zspl[:, 1:-1] = mtilt[:, np.where(aduse)[0]]
+    # zspl[:, 1:-1] = polytilts[arcdet[np.where(aduse)[0]], :].T
+    zspl[:, 0] = zspl[:, 1] + polytilts[0, :] - polytilts[arcdet[np.where(aduse)[0][0]], :]
+    zspl[:, -1] = zspl[:, -2] + polytilts[-1, :] - polytilts[arcdet[np.where(aduse)[0][-1]], :]
+    # Make sure the endpoints are set to 0.0 and 1.0
+    zspl[:, 0] -= zspl[ordcen[0, 0], slit]
+    zspl[:, -1] = zspl[:, -1] - zspl[ordcen[-1, slit], -1] + 1.0
+    # Prepare the spline variables
+    # if False:
+    #     pmin = 0
+    #     pmax = -1
+    # else:
+    #     pmin = int(max(0, np.min(slf._lordloc[det-1])))
+    #     pmax = int(min(msarc.shape[1], np.max(slf._rordloc[det-1])))
+    # xsbs = np.outer(xspl, np.ones(yspl.size))
+    # ysbs = np.outer(np.ones(xspl.size), yspl)
+    # zsbs = zspl[wgd]
+    # Restrict to good portion of the image
+    tiltspl = interpolate.RectBivariateSpline(xspl, yspl, zspl, kx=3, ky=3)
+    yval = np.linspace(0.0, 1.0, msarc.shape[0])
+    tilts = tiltspl(xspl, yval, grid=True).T
+    # Return
+    return tilts
+
+
+def tilts_spline(all_tilts, arcdet, aduse, polytilts, msarc, use_mtilt=False, maskval=-999999.9):
+    msgs.info("Performing a spline fit to the tilts")
+    # Unpack
+    xtilt, ytilt, ztilt, mtilt, wtilt = all_tilts
+    #
+    wgd = np.where((ytilt != maskval) & (ztilt != maskval))
+    txsbs = xtilt[wgd]
+    tysbs = ytilt[wgd]
+    if use_mtilt:
+        tzsbs = mtilt[wgd]/(msarc.shape[0]-1)
+    else:
+        tzsbs = ztilt[wgd]
+    twsbs = wtilt[wgd]
+
+    # Append the end points
+    wlo = np.where((ytilt == np.min(tysbs)) & (ytilt != maskval) & (ztilt != maskval))
+    whi = np.where((ytilt == np.max(tysbs)) & (ytilt != maskval) & (ztilt != maskval))
+    xlo = (xtilt[wlo] * (msarc.shape[1] - 1.0)).astype(np.int)
+    xhi = (xtilt[whi] * (msarc.shape[1] - 1.0)).astype(np.int)
+    xsbs = np.append(xtilt[wlo], np.append(txsbs, xtilt[whi]))
+    ysbs = np.append(np.zeros(wlo[0].size), np.append(tysbs, np.ones(whi[0].size)))
+
+    zlo = ztilt[wlo] + polytilts[0, xlo] - polytilts[arcdet[np.where(aduse)[0][0]], xlo]
+    zhi = ztilt[whi] + polytilts[-1, xhi] - polytilts[arcdet[np.where(aduse)[0][-1]], xhi]
+    zsbs = np.append(zlo, np.append(tzsbs, zhi))
+    wsbs = np.append(wtilt[wlo], np.append(twsbs, wtilt[whi]))
+    # Generate the spline curve
+    tiltspl = interpolate.SmoothBivariateSpline(xsbs, zsbs, ysbs, w=wsbs, kx=3, ky=3,
+                                                s=xsbs.size)#, bbox=[0.0, 1.0, min(zsbs.min(),0.0), max(zsbs.max(),1.0)])
+    xspl = np.linspace(0.0, 1.0, msarc.shape[1])
+    yspl = np.linspace(0.0, 1.0, msarc.shape[0])
+    tilts = tiltspl(xspl, yspl, grid=True).T
+    #
+    tmp = (msarc.shape[0]-1)*tiltspl(np.arange(1490, 1499)/(msarc.shape[1]-1),
```

Comment:
I'm not sure the code ever comes here at the moment (it doesn't seem to), but should we remove this, or are you still debugging?

Reply:
removed